Simplify launch arguments

jonathanKingston opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.

⚠️ If you're an external contributor, please file an issue first before working on a PR, as we can't guarantee that we will accept your changes if they haven't been discussed ahead of time. Thanks!
-->

Task/Issue URL: https://app.asana.com/0/72649045549333/1208825639115711
Tech Design URL:
CC:

**Description**:

Simplifies launch arguments logic so user preferences can be used instead of environment variables.

<!--
If at any point it isn't actively being worked on/ready for review/otherwise moving forward strongly consider closing it (or not opening it in the first place). If you decide not to close it, use Draft PR while work is still in progress or use `DO NOT MERGE` label to clarify the PRs state and comment with more information.
-->

**Steps to test this PR**:
1. Ensure you can skip dialogs by enabling the environment variable
2. Ensure dialogs are present when they're not set.

<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable.
-->

**Definition of Done (Internal Only)**:

* [ ] Does this PR satisfy our [Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f)?

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 15
* [ ] iOS 16
* [ ] iOS 17

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

jonathanKingston committed:
Drop isUITesting arg

jonathanKingston committed:
Make ONBOARDING and isOnboardingComplete control both onboarding and dax dialogs

jonathanKingston committed:
Remove dax dialogs as not needed

jonathanKingston committed:
Add daxdialogs check back

jonathanKingston committed:
Merge remote-tracking branch 'origin/main' into jkt/launch-arguments

alessandroboron commented on DuckDuckGo/LaunchOptionsHandler.swift:
```diff
@@ -32,12 +31,8 @@ public final class LaunchOptionsHandler {
-
     public var isOnboardingCompleted: Bool {
-        userDefaults.string(forKey: Self.isOnboardingcompleted) == "true"
+        ProcessInfo.processInfo.environment["ONBOARDING"] == "false" || userDefaults.string(forKey: Self.isOnboardingcompleted) == "true"
```
We should inject `ProcessInfo.processInfo.environment` so we can test the logic.

`public init(launchArguments: [String] = ProcessInfo.processInfo.arguments, environment: [String: String] =  ProcessInfo.processInfo.environment, userDefaults: UserDefaults = .app) {
   self.launchArguments = launchArguments
   self.environment = environment
   self.userDefaults = userDefaults
}

```suggestion
        environment["ONBOARDING"] == "false" || userDefaults.string(forKey: Self.isOnboardingcompleted) == "true"
```

alessandroboron commented on DuckDuckGoTests/LaunchOptionsHandlerTests.swift:
```diff
@@ -35,31 +35,6 @@ final class LaunchOptionsHandlerTests: XCTestCase {
         try super.tearDownWithError()
     }
 
-    // MARK: - isUITesting
```
We can add tests to ensure that when environment["ONBOARDING"] is true/false then` isOnboardingCompleted` should return `false/true`

alessandroboron commented on DuckDuckGo/MainViewController.swift:
```diff
@@ -508,15 +508,12 @@ class MainViewController: UIViewController {
-        let showOnboarding = !tutorialSettings.hasSeenOnboarding ||
-            // explicitly show onboarding, can be set in the scheme > Run > Environment Variables
-            ProcessInfo.processInfo.environment["ONBOARDING"] == "true"
+        let showOnboarding = !tutorialSettings.hasSeenOnboarding
```
should this be `let showOnboarding = !tutorialSettings.hasSeenOnboarding || !LaunchOptionsHandler().isOnboardingCompleted`?

I think in the previous behaviour the developer could bypass `!tutorialSettings.hasSeenOnboarding` by setting  `ProcessInfo.processInfo.environment["ONBOARDING"] == “true”`. This would force the onboarding to appear without deleting the app.  Now we just rely on `!tutorialSettings.hasSeenOnboarding`. That means that we need to delete the App to see the onboarding again.

alessandroboron commented on DuckDuckGo/DaxDialogs.swift:
```diff
@@ -278,8 +278,9 @@ final class DaxDialogs: NewTabDialogSpecProvider, ContextualOnboardingLogic {
     var isEnabled: Bool {
-        // skip dax dialogs in integration tests
-        guard ProcessInfo.processInfo.environment["DAXDIALOGS"] != "false" else { return false }
+        if LaunchOptionsHandler().isOnboardingCompleted {
```
I know this wasn’t injected before, but do you think we can inject it and test it?

alessandroboron reviewed (commented):
@jonathanKingston thanks for tackling this. The code looks much cleaner now. I left few comments about injection and tests. Let me know if you want to discuss more.

jonathanKingston commented:
Created: https://github.com/duckduckgo/apple-browsers/pull/34 in the unified repo and addressed feedback there. Waiting for CI :). I'll close this for now as we'll be moving there.
